File: U_A_S_D/src/SORTIN_ALGORYTHMS/QUICK_SORT_183_02/QuickSortMethods.java
package SORTIN_ALGORYTHMS.QUICK_SORT_183_02;

public class QuickSortMethods {

    public static void QuickSort(int[] ARR,int low,int high){
        
        if(low < high){
            
            int PivotIndex = Partition(ARR,low,high); // Index of pivot element
            
            QuickSort(ARR,low, PivotIndex-1); // Recursively call for left part
            QuickSort(ARR,PivotIndex+1, high); // Recursively call for right part
        }

    }

    private static int Partition(int[] ARR, int low, int high) {

        int i = (low-1); // Index of current element which we have to swap
        int pivot = ARR[high];

        for (int j = low; j < high; j++) {

            if(ARR[j] < pivot){
                i++;
                int TEMP;
                // Swap if the current element is smaller than pivot
                TEMP = ARR[i];
                ARR[i] = ARR[j];
                ARR[j] = TEMP;
            }
        }

        int TEMP;
        // Swap the pivot element to its correct pisition
        TEMP = ARR[i+1];
        ARR[i+1] = ARR[high];
        ARR[high] = TEMP;

        return i + 1;
    }

}

File: U_A_S_D/src/SORTIN_ALGORYTHMS/QUICK_SORT_183_02_KOTLIN/QuickSortMethods.kt
package SORTIN_ALGORYTHMS.QUICK_SORT_183_02_KOTLIN


// Function Quick Sort ///////////////////////////////////////////////////////////////
    fun quickSort(arr: IntArray, low: Int, high: Int) {

        if (low < high) {
            val pivotIndex = partition(arr, low, high) // Index of pivot element

            quickSort(arr, low, pivotIndex - 1) // Recursively call for left part
            quickSort(arr, pivotIndex + 1, high) // Recursively call for right part
        }
    }


// Function Partition //////////////////////////////////////////////////////////////////
    fun partition(arr: IntArray, low: Int, high: Int): Int {

        var i = low - 1 // Index of current element which we have to swap
        val pivot = arr[high]

        for (j in low until high) {

            if (arr[j] < pivot) {
                i++

                // Swap if the current element is smaller than pivot
                var temp: Int = arr[i]
                arr[i] = arr[j]
                arr[j] = temp
            }
        }

        // Moves the pivot element to its correct position
        val temp: Int = arr[i + 1]
        arr[i + 1] = arr[high]
        arr[high] = temp

        return i + 1
    }
